posts/2017-02-13-more-types-than-classes.md

```diff
@@ -0,0 +1,702 @@
+  on values of a type parameter is a leak in this abstraction. This
+  includes testing the value’s class, converting to string, and
+  comparing to other values of supposedly utter mystery for
+  equality. The ability to create a “default” value is also a leak. A
+  function is always permitted to ask only that of these that it needs
+  from the caller; make them default, and this design choice is taken
+  away. That is why <code>Object#equals</code> is little better for
+  type-safety than reflection-based calls, and why total type erasure
+  is a desirable feature rather than a design flaw—plugging these
+  leaks gives the programmer as much freedom to abstract by
+  information-hiding as she wishes.
+</div>
+
+### How many calls are there?
+
+Put another way, when implementing the code in the scope of a type
+parameter, your implementation must be equally valid _for all_
+possible `G` substitutions, including the ones that haven’t been
+invented yet. This is why we call it _universal_ quantification.
+
+But it is not merely each declaration of a type parameter that yields
+a distinct type—each call does! Consider two consecutive calls to
+`pickGreeting`.
+
+```scala
+pickGreeting(greeting, 42)
+pickGreeting(33, 84)
+```
+
+Externally, there are two `G` types. However, the possibility of
+writing this demands another level of uniqueness treatment when
+typechecking `pickGreeting`’s definition: whatever `G` is now, like
+`String`, it might be something else in the next call, like `Int` in
+the above example. With recursion, it might even be two different
+things at the same time. There’s nothing to hold this at two, either:
+there may be an unbounded number of substitutions for a given type
+parameter within a single program, at a single point in time.
+
+While `G` _may_ be the same between two invocations of `pickGreeting`,
+it might not. So we have no choice but to treat the `G` types of _each
+call_ as separate types. There may be infinitely many calls, so there
+are so many types.
+
+Incidentally, the same happens for singleton types. Each time `val
+greeting` comes into scope, it induces a separate singleton type. It
+is easy enough to arrange for an unbounded number of scope entries in
+a particular program. This isn’t so practical as the type parameter
+phenomenon, though.
+
+## More types from variable copies
+
+Suppose we’d like to wait a while to compute our greeting. We can
+define a type-and-class to represent that conveniently.
+
+```scala
+// like Coyoneda Id, if that helps
+sealed abstract class Later[A] {
+  type I
+  val i: I
+  val f: I => A
+}
+
+def later[In, A](now: In)(later: In => A)
+  : Later[A] = new Later[A] {
+    type I = In
+    val i = now
+    val f = later
+}
+
+val greeting3 = later(3){
+  n => List.fill(n)("hi").mkString(" ")
+}
+```
+
+How many _classes_ are involved here, in the type of `greeting3`?
+
+1. `Later`, obviously;
+2. `String`, the output type of `greeting3.f`;
+3. `Int`, the `I` type.
```

Comment:
Again this might be overly pedantic, but you might also want to mention the `Function1` class, which is involved in the type of `greeting3.f` as well.